### PR TITLE
add heap and stack obj var and fix CI

### PR DESCRIPTION
**Pull request by jumormt**

### CODE REVIEW

**Comment by jumormt** [comment]
@@ -14,7 +14,7 @@
 FunctionObjs        4948
 GepsNum             84876
 GlobalObjs          1147
-HeapObjs            278
+HeapObjs            279
 IndCallSites        1132
 LoadsNum            52473
 MaxStructSize       17347
@@ -31,9 +31,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.221
-SVFIRTime           2.466
-SymbolTableTime     0.556
+LLVMIRTime          4.237
+SVFIRTime           2.485
+SymbolTableTime     0.565
 #######################################################

 *********PTACallGraph Stats (Andersen analysis)***************
@@ -60,11 +60,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16479e+06
-MemoryUsageVmsize   4.16454e+06
+MemoryUsageVmrss    4.1647e+06
+MemoryUsageVmsize   4.16455e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           125.786
+TotalTime           128.365
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
@@ -115,32 +115,32 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673055
+TotalUnions              673046
 PropertyUnions           222658
-UniqueUnions             19262
-LookupUnions             395146
-PreemptiveUnions         35989
+UniqueUnions             19260
+LookupUnions             395148
+PreemptiveUnions         35980
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182804
-PropertyIntersections    4145581
-UniqueIntersections      776
-LookupIntersections      2157
-PreemptiveIntersections  34290
+TotalIntersections       4182797
+PropertyIntersections    4145547
+UniqueIntersections      826
+LookupIntersections      2141
+PreemptiveIntersections  34283
 #######################################################

 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
 AverageRegSize      85.727
-GenMUCHITime        1.85
-GenRegionTime       471.468
-InsertPHITime       0.567
-SSARenameTime       0.086
-TotalMSSATime       473.978
+GenMUCHITime        1.925
+GenRegionTime       508.023
+InsertPHITime       0.597
+SSARenameTime       0.082
+TotalMSSATime       510.63
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
 CSChiNode           111246
@@ -166,10 +166,10 @@
 ATNodeTime          0.843
 AvgWeight           167.789
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.437
+ConnIndEdgeTime     4.413
 OptTime             0
 TLNodeTime          0
-TotalTime           5.28
+TotalTime           5.256
 ----------------Numbers stats----------------------------
 ActualIn            163193
 ActualOut           111246
@@ -231,20 +231,20 @@
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.56573e+06
-MemoryUsageVmsize   3.59792e+06
+MemoryUsageVmrss    3.58227e+06
+MemoryUsageVmsize   3.6171e+06
 PhiTime             0
-PrelabelingTime     0.146
+PrelabelingTime     0.138
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           422.534
+SolveTime           399.004
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           631.834
+TotalTime           611.892
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    205.968
+meldLabelingTime    209.504
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
@@ -262,13 +262,13 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       157897
+ProcessedCopy       162576
 ProcessedFRet       0
-ProcessedGep        849351
-ProcessedLoad       860526
+ProcessedGep        867537
+ProcessedLoad       874776
 ProcessedMSSANode   2293638
-ProcessedPhi        101413
-ProcessedStore      317390
+ProcessedPhi        104067
+ProcessedStore      320425
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
@@ -283,12 +283,12 @@

 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29099
-TotalUnions              1108027036
-PropertyUnions           972530933
-UniqueUnions             35555
-LookupUnions             135412411
-PreemptiveUnions         48137
+UniquePointsToSets       29724
+TotalUnions              1142383129
+PropertyUnions           987717820
+UniqueUnions             38097
+LookupUnions             154575935
+PreemptiveUnions         51277
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0

File: svf/include/DDA/DDAVFSolver.h [hunk 1]
```diff
@@ -473,7 +473,7 @@ class DDAVFSolver
         NodeID id = getPtrNodeID(var);
         const MemObj* obj = _pag->getObject(id);
```
Comment (yuleisui) [inline svf/include/DDA/DDAVFSolver.h:474]: baseobjvar = _pag->getObject(id);
isa<Stack>(baseobjvar)

File: svf/include/DDA/DDAVFSolver.h [hunk 2]
```diff
@@ -637,9 +637,7 @@ class DDAVFSolver
     //@{
     virtual inline bool isHeapCondMemObj(const CVar& var, const StoreSVFGNode*)
     {
-        const MemObj* mem = _pag->getObject(getPtrNodeID(var));
-        assert(mem && "memory object is null??");
-        return mem->isHeap();
+        return SVFUtil::isHeapOriginVar(_pag->getGNode(getPtrNodeID(var)));
```
Comment (yuleisui) [inline svf/include/DDA/DDAVFSolver.h:640]: remove isHeapOriginVar.

File: svf/include/Util/SVFUtil.h [hunk 3]
```diff
@@ -374,7 +374,11 @@ bool isHeapAllocExtCallViaRet(const CallICFGNode* cs);
 
 bool isHeapAllocExtCall(const ICFGNode* cs);
 
+/// Check if the given variable originates from the heap.
+bool isHeapOriginVar(const SVFVar* var);
```
Comment (yuleisui) [inline svf/include/Util/SVFUtil.h:378]: remove

File: svf/include/Util/SVFUtil.h [hunk 4]
```diff
@@ -374,7 +374,11 @@ bool isHeapAllocExtCallViaRet(const CallICFGNode* cs);
...
 
+/// Check if the given variable originates from the heap.
+bool isStackOriginVar(const SVFVar* var);
```
Comment (yuleisui) [inline svf/include/Util/SVFUtil.h:381]: remove

File: svf/lib/MSSA/MemRegion.cpp [hunk 5]
```diff
@@ -577,13 +577,15 @@ bool MRGenerator::isNonLocalObject(NodeID id, const SVFFunction* curFun) const
     const MemObj* obj = pta->getPAG()->getObject(id);
```
Comment (yuleisui) [inline svf/lib/MSSA/MemRegion.cpp:577]: getObject => getBaseObj.

**Comment by jumormt** [comment]
--- log/redis-server.log        2024-12-16 17:53:12.879406997 +1100
+++ log/redis-server-xiao.log   2024-12-16 17:54:56.614375761 +1100
@@ -14,7 +14,7 @@
 FunctionObjs        4948
 GepsNum             84876
 GlobalObjs          1147
-HeapObjs            278
+HeapObjs            279
 IndCallSites        1132
 LoadsNum            52473
 MaxStructSize       17347
@@ -31,9 +31,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.514
-SVFIRTime           2.584
-SymbolTableTime     0.583
+LLVMIRTime          4.482
+SVFIRTime           2.605
+SymbolTableTime     0.581
 #######################################################

 *********PTACallGraph Stats (Andersen analysis)***************
@@ -60,11 +60,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16498e+06
-MemoryUsageVmsize   4.16455e+06
+MemoryUsageVmrss    4.16468e+06
+MemoryUsageVmsize   4.16452e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           128.157
+TotalTime           127.262
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
@@ -115,47 +115,47 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673061
+TotalUnions              673050
 PropertyUnions           222658
-UniqueUnions             19262
-LookupUnions             395146
-PreemptiveUnions         35995
+UniqueUnions             19260
+LookupUnions             395148
+PreemptiveUnions         35984
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182810
-PropertyIntersections    4145482
-UniqueIntersections      845
-LookupIntersections      2187
-PreemptiveIntersections  34296
+TotalIntersections       4182801
+PropertyIntersections    4145547
+UniqueIntersections      835
+LookupIntersections      2132
+PreemptiveIntersections  34287
 #######################################################

 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-AverageRegSize      85.4739
-GenMUCHITime        1.941
-GenRegionTime       462.345
-InsertPHITime       0.554
-SSARenameTime       0.108
-TotalMSSATime       464.956
+AverageRegSize      85.727
+GenMUCHITime        1.841
+GenRegionTime       554.86
+InsertPHITime       0.546
+SSARenameTime       0.088
+TotalMSSATime       557.346
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
-CSChiNode           112085
+CSChiNode           111246
 CSHasChi            22679
 CSHasMu             24549
-CSMuNode            164169
-FunEntryChi         24443
+CSMuNode            163193
+FunEntryChi         24294
 FunHasEntryChi      3967
 FunHasRetMu         4630
-FunRetMu            24374
+FunRetMu            24225
 LoadHasMu           24342
 LoadMuNode          43202
-MSSAPhi             59993
+MSSAPhi             59315
 MaxRegSize          1650
-MemRegions          4556
+MemRegions          4538
 StoreChiNode        14269
 StoreHasChi         10507
 #######################################################
@@ -163,16 +163,16 @@
 *********SVFG Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-ATNodeTime          0.849
-AvgWeight           166.605
+ATNodeTime          0.855
+AvgWeight           167.789
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.427
+ConnIndEdgeTime     4.447
 OptTime             0
 TLNodeTime          0
-TotalTime           5.276
+TotalTime           5.302
 ----------------Numbers stats----------------------------
-ActualIn            164169
-ActualOut           112085
+ActualIn            163193
+ActualOut           111246
 ActualParam         57426
 ActualRet           7530
 Addr                17887
@@ -184,25 +184,25 @@
 DirectCallEdge      41434
 DirectEdge          292884
 DirectRetEdge       7351
-FormalIn            24443
-FormalOut           24374
+FormalIn            24294
+FormalOut           24225
 FormalParam         7869
 FormalRet           795
 Gep                 84876
-IndCallEdge         191215
-IndRetEdge          125679
-IndirectEdge        722123
-IndirectEdgeLabels  120309251
+IndCallEdge         189998
+IndRetEdge          124671
+IndirectEdge        717043
+IndirectEdgeLabels  120311897
 Load                24342
-MSSAPhi             59993
+MSSAPhi             59315
 MaxInDegree         9710
 MaxIndInDeg         9710
 MaxIndOutDeg        2458
 MaxOutDegree        9600
 PHI                 6776
 Store               24164
-TotalEdge           1015007
-TotalNode           630868
+TotalEdge           1009927
+TotalNode           628077
 #######################################################

 *********PTACallGraph Stats (Flow-sensitive analysis)***************
@@ -223,7 +223,7 @@
 ----------------Time and memory stats--------------------
 AddrTime            0
 AverageSCCSize      19.105
-AvgPtsSize          67.7949
+AvgPtsSize          67.7948
 AvgTopLvlPtsSize    369.968
 AvgVersionPtsSize   32.3067
 CopyTime            0
@@ -231,20 +231,20 @@
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.56773e+06
-MemoryUsageVmsize   3.6026e+06
+MemoryUsageVmrss    3.5817e+06
+MemoryUsageVmsize   3.61182e+06
 PhiTime             0
-PrelabelingTime     0.149
+PrelabelingTime     0.152
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           413.842
+SolveTime           424.95
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           625.016
+TotalTime           627.456
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    207.838
+meldLabelingTime    209.153
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
@@ -262,33 +262,33 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       158384
+ProcessedCopy       161192
 ProcessedFRet       0
-ProcessedGep        847425
-ProcessedLoad       837199
-ProcessedMSSANode   2310384
-ProcessedPhi        101646
-ProcessedStore      313925
+ProcessedGep        859709
+ProcessedLoad       856043
+ProcessedMSSANode   2293638
+ProcessedPhi        103499
+ProcessedStore      317670
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
 TotalEmptyVPts      0
-TotalExistingVPts   15080518
-TotalNonEmptyVPts   15080518
+TotalExistingVPts   15080539
+TotalNonEmptyVPts   15080539
 TotalObjects        15741
 TotalPointers       497044
 TotalSingleVObjs    0
-TotalVersions       17442432
+TotalVersions       17442525
 #######################################################

 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29203
-TotalUnions              1095942141
-PropertyUnions           958021873
-UniqueUnions             36016
-LookupUnions             137835599
-PreemptiveUnions         48653
+UniquePointsToSets       29493
+TotalUnions              1119345511
+PropertyUnions           965154842
+UniqueUnions             37583
+LookupUnions             154102445
+PreemptiveUnions         50641
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0

**Pull request merged**